### PR TITLE
Add support for ICMPv6 to ICMP sockets

### PR DESCRIPTION
**Pull request by progval**
Hi,

I started adding support for ICMPv6 to ICMP sockets (see #99 and #103).

Do you have some feedback on my changes (attached to this PR) before I implement it further?

Thanks!

### CODE REVIEW

File: src/wire/mod.rs [hunk 1]
```diff
@@ -174,6 +176,13 @@ pub use self::icmpv6::{Message as Icmpv6Message,
                        ParamProblem as Icmpv6ParamProblem,
                        Packet as Icmpv6Packet,
                        Repr as Icmpv6Repr};
+#[cfg(any(feature = "proto-ipv4", feature = "proto-ipv6"))]
+pub use self::icmp::{Message as IcmpMessage,
+                     DstUnreachable as IcmpDstUnreachable,
+                     TimeExceeded as IcmpTimeExceeded,
+                     ParamProblem as IcmpParamProblem,
+                     Packet as IcmpPacket,
+                     Repr as IcmpRepr};
```
Comment (whitequark) [inline src/wire/mod.rs]: Are you planning to somehow unify the ICMPv4 and ICMPv6 `Message`, `DstUnreachable`, `TimeExceeded` and `ParamProblem` enums? This doesn't seem like the right way to go to me. Why is this necessary?

**Comment by whitequark** [comment]
Other than the comment above, looks good!

File: src/iface/ethernet.rs [hunk 2]
```diff
@@ -826,13 +831,13 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
         packet
     }
 
-    #[cfg(feature = "proto-ipv4")]
-    fn process_icmpv4<'frame>(&self, _sockets: &mut SocketSet, ip_repr: IpRepr,
-                              ip_payload: &'frame [u8]) -> Result<Packet<'frame>>
+    #[cfg(any(feature = "proto-ipv4", feature = "proto-ipv6"))]
+    fn process_icmp<'frame>(&self, _sockets: &mut SocketSet, ip_repr: IpRepr,
```
Comment (dlrobertson) [inline src/iface/ethernet.rs]: ICMPv6 packets will be processed by [`process_icmpv6`](https://github.com/m-labs/smoltcp/blob/master/src/iface/ethernet.rs#L741). It has a different [IP protocol number](https://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml).

File: src/wire/mod.rs [hunk 3]
```diff
@@ -174,6 +176,13 @@ pub use self::icmpv6::{Message as Icmpv6Message,
                        ParamProblem as Icmpv6ParamProblem,
                        Packet as Icmpv6Packet,
                        Repr as Icmpv6Repr};
+#[cfg(any(feature = "proto-ipv4", feature = "proto-ipv6"))]
```
Comment (dlrobertson) [inline src/wire/mod.rs]: Same

File: src/wire/mod.rs [hunk 4]
```diff
@@ -95,6 +95,8 @@ mod ipv6fragment;
 mod icmpv4;
 #[cfg(feature = "proto-ipv6")]
 mod icmpv6;
+#[cfg(all(feature = "proto-ipv4", feature = "proto-ipv6"))]
```
Comment (dlrobertson) [inline src/wire/mod.rs]: Rely on the `icmpv6` module for `proto-ipv6`.

File: src/wire/mod.rs (hunk 1 above)
Comment (ProgVal) [reply src/wire/mod.rs]: Indeed, it isn't

File: src/wire/mod.rs (hunk 4 above)
Comment (dlrobertson) [reply src/wire/mod.rs]: The comment by @whitequark [here](https://github.com/m-labs/smoltcp/pull/205/files#r187041666) was kind of what I was getting at and is a much clearer. If we get an ICMPv6 response we shouldn't be using the definitions in `wire::icmpv4`.

Note: my assumption was that `wire::icmp` was the same thing as `wire::icmpv4`. If your intent was to make `wire::icmp` a module that encapsulates ICMPv4 and ICMPv4 (similar to `wire::ip` for IPv4 and IPv6), then your current approach is correct.

**Comment by whitequark** [comment]
OK, now that I see `icmp.rs`, I think this is definitely not the right approach. `IcmpRepr` makes sense to have, but the other enums aren't used anywhere and it's not clear to me that they are clearly useful as defined.

**Comment by ProgVal** [comment]
Yeah, I started by implementing all the enums before looking at the existing code. I'll remove them

File: src/wire/mod.rs (hunk 4 above)
Comment (ProgVal) [reply src/wire/mod.rs]: Hmm, except that it should be `any()` and not `all()`

Comment (dlrobertson) [reply src/wire/mod.rs]: True.

**Comment by ProgVal** [comment]
I'm getting three test failures:

```
---- socket::icmp::test::test_accepts_udp_v6 stdout ----
	thread 'socket::icmp::test::test_accepts_udp_v6' panicked at 'assertion failed: `(left == right)`
  left: `Err(Truncated)`,
 right: `Ok(())`', src/socket/icmp.rs:745:9

---- socket::icmp::test::test_send_dispatch_v6 stdout ----
	thread 'socket::icmp::test::test_send_dispatch_v6' panicked at 'assertion failed: `(left == right)`
  left: `Err(Checksum)`,
 right: `Err(Unaddressable)`', src/socket/icmp.rs:499:9

---- socket::icmp::test::test_set_hop_limit_v6 stdout ----
	thread 'socket::icmp::test::test_set_hop_limit_v6' panicked at 'assertion failed: `(left == right)`
  left: `Err(Checksum)`,
 right: `Ok(())`', src/socket/icmp.rs:552:9
```

My guess is that the two latter are because the checksum is never computed (which is not an issue for ICMPv4 tests, because ICMPv4 checksums can be blank).

I tried solving them, with no success. Any idea?

**Comment by whitequark** [comment]
> ---- socket::icmp::test::test_accepts_udp_v6 stdout ----
> 	thread 'socket::icmp::test::test_accepts_udp_v6' panicked at 'assertion failed: `(left == right)`
>   left: `Err(Truncated)`,
>  right: `Ok(())`', src/socket/icmp.rs:745:9

Do you know why this results in a Truncated error?

**Comment by ProgVal** [comment]
No. I added a `println!` in front of every initialization of `Truncated` I could find in ICMP- and IPv6-related files, none of them were called.

File: src/iface/ethernet.rs [hunk 5]
```diff
@@ -828,11 +833,11 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
 
     #[cfg(feature = "proto-ipv4")]
     fn process_icmpv4<'frame>(&self, _sockets: &mut SocketSet, ip_repr: IpRepr,
-                              ip_payload: &'frame [u8]) -> Result<Packet<'frame>>
+                            ip_payload: &'frame [u8]) -> Result<Packet<'frame>>
```
Comment (dlrobertson) [inline src/iface/ethernet.rs]: nit: whitespace

File: src/iface/ethernet.rs [hunk 6]
```diff
@@ -828,11 +833,11 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
...
     {
         let icmp_packet = Icmpv4Packet::new_checked(ip_payload)?;
         let checksum_caps = self.device_capabilities.checksum.clone();
-        let icmp_repr = Icmpv4Repr::parse(&icmp_packet, &checksum_caps)?;
+        let icmp_repr = Icmpv4Repr::parse(&icmp_packet, &checksum_caps)?.into();
```
Comment (dlrobertson) [inline src/iface/ethernet.rs]: The new `IcmpRepr` really shouldn't be used in the `process_icmpv4` or `process_icmpv6` functions. AFAIK it is really only needed for getting data from the socket to the interface, but I don't think it is needed for parsing data from the outside world.

File: src/socket/icmp.rs [hunk 7]
```diff
@@ -297,7 +313,20 @@ impl<'a, 'b> IcmpSocket<'a, 'b> {
                         payload_len: repr.buffer_len(),
                         hop_limit:   hop_limit,
                     });
-                    emit((ip_repr, repr))
+                    emit((ip_repr, IcmpRepr::Ipv4(repr)))
+                },
+                IpAddress::Ipv6(ipv6_addr) => {
+                    let packet = Icmpv6Packet::new(&*packet_buf);
+                    let src_addr = Ipv6Address::default();
+                    let repr = Icmpv6Repr::parse(&src_addr.into(), &ipv6_addr.into(), &packet, &ChecksumCapabilities::default())?;
```
Comment (dlrobertson) [inline src/socket/icmp.rs]: Should we expect users to fill in the checksum? Or should we just parse it with `ChecksumCapabilities::ignored()`?

Comment (whitequark) [reply src/socket/icmp.rs]: We shouldn't expect the users to fill in checksum.

Comment (dlrobertson) [reply src/socket/icmp.rs]: @ProgVal you might update the corresponding line for ICMPv4 as well https://github.com/m-labs/smoltcp/pull/205/files#diff-7dff2f1b7b74bbfcd78f956578d349d1R308

Comment (ProgVal) [reply src/socket/icmp.rs]: I don't really understand what `ChecksumCapabilities` mean (should the user of smoltcp verify them? or smoltcp itself? or the hardware?), but using `ChecksumCapabilities::ignored` fixed two of the issues, thanks!

File: src/iface/ethernet.rs (hunk 6 above)
Comment (ProgVal) [reply src/iface/ethernet.rs]: It's needed for `IcmpSocket::send`.

**Comment by ProgVal** [comment]
Ha, I found why it's truncated, it's because the buffer allocated on the first line of `test_accepts_udp_v6` only supports 46-bytes packets.

File: src/iface/ethernet.rs (hunk 6 above)
Comment (dlrobertson) [reply src/iface/ethernet.rs]: yeah this is more or less a nit. It might be cleaner to add a line like `let socket_repr = IcmpRepr::from(icmp_repr)` then use `socket_repr` for the `IcmpSocket` functions and then you can still `match icmp_repr` and you don't have the extra `IcmpRepr::Ipv4(...)`, but this is just a nit.

File: src/socket/icmp.rs (hunk 7 above)
Comment (dlrobertson) [reply src/socket/icmp.rs]: It is essentially used to determine if the parse/emit function should verify/fill the checksum. It is typically used in the context of a device. For example, a device that we can offload IPv4 checksumming on transmit to would have a `ChecksumCapabilities` structure with `ipv4` equal to [TX](https://github.com/m-labs/smoltcp/blob/master/src/phy/mod.rs#L127). However, in this case we don't want to evaluate the checksum so using `ChecksumCapabilities::ignored` makes that happen.

File: src/iface/ethernet.rs (hunk 6 above)
Comment (dlrobertson) [reply src/iface/ethernet.rs]: You also need to add similar code to `process_icmpv6` to ensure any ICMPv6 sockets actually receive data.

File: src/wire/icmp.rs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+                $class::Ipv4(s)
+            }
+        }
+        impl From<icmpv6::$class> for $class {
+            fn from(s: icmpv6::$class) -> Self {
+                $class::Ipv6(s)
+            }
+        }
+    };
+    ( $class:ident < $typearg:tt > ) => {
```
Comment (dlrobertson) [inline src/wire/icmp.rs]: This is the only matching arm that is used right? Or am I missing something?

Comment (ProgVal) [reply src/wire/icmp.rs]: Yes, now it is. I think I'll remove the macro altogether, as there is only one class left.

**Comment by ProgVal** [comment]
(I just squashed some commits to make the git history cleaner)

Looks like everything is working now, thanks for your help!

File: src/socket/icmp.rs [hunk 9]
```diff
@@ -244,13 +248,23 @@ impl<'a, 'b> IcmpSocket<'a, 'b> {
...
+            #[cfg(feature = "proto-ipv4")]
+            (&Endpoint::Udp(endpoint), &IcmpRepr::Ipv4(Icmpv4Repr::DstUnreachable { data, .. }))
+                    if endpoint.addr.is_unspecified() || endpoint.addr == ip_repr.dst_addr() => {
+                let packet = UdpPacket::new(data);
+                match UdpRepr::parse(&packet, &ip_repr.src_addr(), &ip_repr.dst_addr(), cksum) {
+                    Ok(repr) => endpoint.port == repr.src_port,
+                    Err(_) => false,
+                }
+            }
+            #[cfg(feature = "proto-ipv6")]
```
Comment (whitequark) [inline src/socket/icmp.rs]: I feel like these two cases ought to be unified, but not sure if this can be done cleanly.

**Comment by whitequark** [comment]
Looks good! Please squash all commits. After that, r=me.

File: src/socket/icmp.rs (hunk 9 above)
Comment (ProgVal) [reply src/socket/icmp.rs]: I could add a function to factorize the code, but I'm not sure it improves readability

Comment (whitequark) [reply src/socket/icmp.rs]: I was thinking of something like
```rust
match x {
  #[cfg(foo)] Y | #[cfg(bar)] Z => t
}
```
but I don't know if it works.

Comment (ProgVal) [reply src/socket/icmp.rs]: It doesn't, I tried

**Comment by whitequark** [comment]
@m-labs-homu r+

**Comment by m-labs-homu** [comment]
:pushpin: Commit 6eb23fa has been approved by `whitequark`

**Comment by m-labs-homu** [comment]
:hourglass: Testing commit 6eb23fa73302f04ba371ee049c597cac84fa1f66 with merge ce2fbb7dd2a45d27426db3f842a2f5fa68f7b700...

**Comment by dlrobertson** [comment]
@ProgVal thanks for working on this!
